File: 1046-last-stone-weight/1046-last-stone-weight.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        return solve(stones);    
    }
private:
    int solve(vector<int>& stones) {
        priority_queue<int> maxHeap;
        
        for(int stone: stones)
            maxHeap.push(stone);
        
        while(maxHeap.size() > 1) {
            int y = maxHeap.top();
            maxHeap.pop();
            int x = maxHeap.top();
            maxHeap.pop();
            
            if(x != y)
                maxHeap.push(y - x);
        }
        
        if(maxHeap.empty())
            return 0;
        return maxHeap.top();
    }
};

File: 1046-last-stone-weight/1046-last-stone-weight.py
import heapq
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        maxHeap = []
        
        for i in stones:
            heapq.heappush(maxHeap, -i)
        
        while len(maxHeap) > 1:
            y = -heapq.heappop(maxHeap)
            x = -heapq.heappop(maxHeap)
            
            if x != y:
                heapq.heappush(maxHeap, -1 * (y - x))
        
        if len(maxHeap) == 0:
            return 0
        return -maxHeap[0]